### PR TITLE
FIX ProxNewton solver with fixpoint strategy

### LINKED ISSUE #256: BUG ProxNewton with `ws_strategy="fixpoint"` is 100 times slower than with subdiff_dist strategy
```python
import numpy as np
from time import time
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from numpy.linalg import norm

from skglm import GeneralizedLinearEstimator
from skglm import datafits
from skglm import penalties
from skglm.solvers import ProxNewton
from skglm.utils.data import make_correlated_data

X, y, _ = make_correlated_data(500, 500, random_state=0)


datafit = datafits.Quadratic()
penalty = penalties.L1(alpha=1)
alpha_max = penalty.alpha_max(datafit.gradient(X, y, np.zeros(len(y))))

penalty.alpha = alpha_max / 10

for ws_strategy in ["subdiff", "fixpoint"]:

    solver = ProxNewton(verbose=1, max_iter=10, warm_start=True, fit_intercept=False,
                        tol=1e-4, ws_strategy=ws_strategy)

    start_time = time()

    model = GeneralizedLinearEstimator(
        datafit=datafit, penalty=penalty, solver=solver).fit(X, y)
    result = model.coef_
    print(f"fit time : {(time() - start_time):.2e} s")
    print("support size: ", (result !=0).sum() )
```

Rsults in:
```
Iteration 1: 73.8784855691, stopping crit: 3.41e+00
Iteration 2: 56.9243492708, stopping crit: 2.32e+00
Iteration 3: 38.2902646369, stopping crit: 1.48e+00
Iteration 4: 31.6128543717, stopping crit: 3.98e-01
Iteration 5: 30.5363866649, stopping crit: 2.61e-01
Iteration 6: 30.4434200241, stopping crit: 8.15e-03
Iteration 7: 30.4432029343, stopping crit: 2.78e-04
Iteration 8: 30.4432024377, stopping crit: 1.05e-05
Stopping criterion max violation: 1.05e-05
fit time : 3.44e-02 s
support size:  97
Iteration 1: 73.8784855691, stopping crit: 3.40e+00
Iteration 2: 53.2290072274, stopping crit: 2.30e+00
Iteration 3: 37.7015535418, stopping crit: 1.13e+00
Iteration 4: 31.7132828676, stopping crit: 5.23e-01
Iteration 5: 30.6305400410, stopping crit: 3.40e-01
Iteration 6: 30.4670263326, stopping crit: 8.02e-02
Iteration 7: 30.4479250663, stopping crit: 3.28e-02
Iteration 8: 30.4443919512, stopping crit: 2.33e-02
Iteration 9: 30.4436981339, stopping crit: 1.79e-02
Iteration 10: 30.4434383185, stopping crit: 1.23e-02
/home/mathurin/workspace/skglm/skglm/solvers/prox_newton.py:188: ConvergenceWarning: `ProxNewton` did not converge for tol=1.000e-04 and max_iter=10.
Consider increasing `max_iter` and/or `tol`.
  warnings.warn(
fit time : 6.81e+00 s
support size:  97
```
so the same solution is found, but it takes 200x more time with `fixpoint`

### PR DESCRIPTION
**Pull request by mathurinm**
fixes #256 

Hard to test properly, but the following now works fine:
```python
import numpy as np
import matplotlib.pyplot as plt
from sklearn.model_selection import train_test_split
from numpy.linalg import norm
from skglm.utils.jit_compilation import compiled_clone

from skglm import GeneralizedLinearEstimator
from skglm import datafits
from skglm import penalties
from skglm.solvers import ProxNewton
from skglm.utils.data import make_correlated_data

X, y, _ = make_correlated_data(500, 5000, random_state=0)

y = np.abs(y) // 1

# datafit = compiled_clone(datafits.Quadratic())
datafit = compiled_clone(datafits.Poisson())
penalty = compiled_clone(penalties.L1(alpha=1))
alpha_max = penalty.alpha_max(datafit.gradient(X, y, np.zeros(len(y))))

penalty.alpha = alpha_max / 10

solver = ProxNewton(verbose=3, max_iter=20, warm_start=True, fit_intercept=False, tol=1e-4, ws_strategy="fixedpoint", max_pn_iter=20)

solver.solve(X, y, datafit, penalty)
```

### CODE REVIEW

File: skglm/solvers/anderson_cd.py [hunk 1]
```diff
@@ -184,7 +184,7 @@ def solve(self, X, y, datafit, penalty, w_init=None, Xw_init=None):
                         opt_ws = penalty.subdiff_distance(w[:n_features], grad_ws, ws)
                     elif self.ws_strategy == "fixpoint":
                         opt_ws = dist_fix_point_cd(
-                            w[:n_features], grad_ws, lipschitz, datafit, penalty, ws
+                            w[:n_features], grad_ws, lipschitz[ws], datafit, penalty, ws
```
Comment (Badr-MOUFAD) [inline skglm/solvers/anderson_cd.py:187]: I feel that we should expect performance regression in ``AnderonCD`` as after this we will be making copies

WDYT @mathurinm?

**Comment by mathurinm** [comment]
Frankly compared to the cost of the rest of the computation I expect this
to be totally negligible

Il sab 1 giu 2024, 12:15 Badr MOUFAD ***@***.***> ha scritto:

File: skglm/solvers/prox_newton.py [hunk 2]
```diff
@@ -253,7 +256,7 @@ def _descent_direction(X, y, w_epoch, Xw_epoch, fit_intercept, grad_ws, datafit,
                 opt = penalty.subdiff_distance(current_w, past_grads, ws)
             elif ws_strategy == "fixpoint":
                 opt = dist_fix_point_cd(
-                    current_w, past_grads, lipschitz, datafit, penalty, ws
+                    current_w, past_grads, lipschitz[ws], datafit, penalty, ws
```
Comment (Badr-MOUFAD) [inline skglm/solvers/prox_newton.py:259]:
```suggestion
                    current_w, past_grads, lipschitz, datafit, penalty, ws
```

lipschitz has already the size of `len(ws)`

**Review by Badr-MOUFAD — COMMENTED** [review]
We have also

https://github.com/scikit-learn-contrib/skglm/blob/96826609a06b9b8cb78382177d1122f2911eede0/skglm/solvers/multitask_bcd.py#L234

We better adopt the same convention here as well.

I'm +1 with tackling it in this PR as this it touches the fixpoint ws strategy.

File: skglm/solvers/prox_newton.py (hunk 2 above)
Comment (mathurinm) [reply skglm/solvers/prox_newton.py]: Good catch ; I renamed the variables to `lipschitz_ws` for clarity

**Comment by mathurinm** [comment]
@Badr-MOUFAD merge if happy

**Comment by mathurinm** [comment]
@Badr-MOUFAD With `L1` it works fine, but with L0_5 the solver is still stuck. 
If I disable the working set (by passing `p0 = n_features` for solver), it works fine.

```python
import numpy as np
from skglm.utils.jit_compilation import compiled_clone

from skglm import datafits
from skglm import penalties
from skglm.solvers import ProxNewton
from skglm.utils.data import make_correlated_data

X, y, _ = make_correlated_data(50, 100, random_state=0)

y = np.abs(y) // 1

datafit = compiled_clone(datafits.Quadratic())
penalty = compiled_clone(penalties.L0_5(alpha=1))
# penalty = compiled_clone(penalties.L1(alpha=1))
alpha_max = penalties.L1(alpha=1).alpha_max(datafit.gradient(X, y, np.zeros(len(y))))

penalty.alpha = alpha_max / 10


solver = ProxNewton(verbose=3, max_iter=20, warm_start=True, fit_intercept=False, tol=1e-4, ws_strategy="fixpoint", max_pn_iter=20, p0=10)
solver.solve(X, y, datafit, penalty)
```

**Review by Badr-MOUFAD — APPROVED** [review]
I feel that the issue related to the `L_05` penalty is not closely related to this PR and suggest investigating it later in another PR.

Thanks for the fix @mathurinm 🚀

**Pull request merged**